### PR TITLE
Add prompt to save scripts on quit [4.0]

### LINKED ISSUE #55026: Unsaved scripts don't prompt when closing editor
### Godot version

098e3cf / 3.4

### System information

W10

### Issue description

When you modify a script, it gets the "modified" status, but it's ignored by the editor. When you quit, any script changes are lost. Only unsaved scenes are handled.

Resolving this would effectively resolve #36281 (after #54653)

### Steps to reproduce

1. Open any script in script editor
2. Modify it
3. Close Godot

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by ator-dev**
Fix #55026

### CODE REVIEW

File: editor/plugins/script_editor_plugin.cpp [hunk 1]
```diff
@@ -2439,6 +2439,10 @@ void ScriptEditor::save_current_script() {
 }
 
 void ScriptEditor::save_all_scripts() {
+	if (close_without_saving) {
+		return; // TODO: Find a better way of preventing scripts from saving on quit.
+	}
```
Comment (KoBeWi) [inline editor/plugins/script_editor_plugin.cpp:2444]: Why is this needed? `save_all_scripts()` isn't normally called when exiting.

Comment (ator-dev) [reply editor/plugins/script_editor_plugin.cpp:2444]: This confused me as well, but I did a lot of tests and scripts always seemed to be saved... It's also in the engine logic; there's a loop that goes through all plugins (including script editor) and saves their external data.
If you could help me work out why, that would be greatly appreciated as I really don't like my solution (;

Comment (KoBeWi) [reply editor/plugins/script_editor_plugin.cpp:2444]: I investigated it and it doesn't make any sense xd
This method is called from EditorNode using `save_external_data()`. It's called when you quit the editor, but it doesn't actually save anything. I debugged it, it goes correctly over all lines and even saves scenes when saving builtin-scripts. But when you reopen the project, no script nor any scene is actually saved.
wtf

This is only on vanilla build. With your changes, when you close with Don't Save, this method actually has some effect.

Comment (ator-dev) [reply editor/plugins/script_editor_plugin.cpp:2444]: Wow, freaky... I did have a go at debugging it but didn't realise it was working that weirdly lol
I'm pretty new to all of this, but I'm guessing one of us should look into it and possibly open a pull request (or modify this one) if we can work out what needs fixing, if anything?

**Review by Paulb23 — COMMENTED** [review]
Seems okay, couple comments inline, though I do wonder about ordering, whether we should save the scenes first, then the scripts.

Also whether we need to do something similar for the shader editor.

File: editor/plugins/script_editor_plugin.h [hunk 2]
```diff
@@ -469,6 +469,8 @@ class ScriptEditor : public PanelContainer {
 	static void _bind_methods();
 
 public:
+	bool close_without_saving; // TODO: Find a better way of preventing scripts from saving on quit.
```
[the remark this answers is not in the document]
Comment (ator-dev) [reply editor/plugins/script_editor_plugin.h:472]: Thanks for the advice - in my new version (which I'll submit after some more testing) I think I've managed to eliminate this, but if I need to re-add it I'll make sure to use a setter.

File: editor/editor_node.cpp [hunk 3]
```diff
@@ -2792,14 +2792,38 @@ void EditorNode::_menu_option_confirm(int p_option, bool p_confirmed) {
...
+
+					save_confirmation->get_ok_button()->set_text(TTR("Save & Quit"));
+					save_confirmation->set_text((p_option == FILE_QUIT ? TTR("Save changes to the following script(s) before quitting?") : TTR("Save changes to the following script(s) before opening Project Manager?")) + unsaved_scripts);
+					save_confirmation->popup_centered();
+				}
+				break;
+			}
+
+			// User has chosen to save scripts before quit.
+			ScriptEditor::get_singleton()->save_all_scripts();
```
[the remark this answers is not in the document]
Comment (ator-dev) [reply editor/editor_node.cpp:2814]: You're right, that's a better, more abstracted way of doing it. Here's my replacement:
image

**Review by Paulb23 — COMMENTED** [review]
I don't think there's necessarily a right order as it could be argued that changing a scene involves it's resources. I was thinking here if we could avoid additional popups. Though thinking about it, that would only apply to built-in scripts?

**Comment by ator-dev** [comment]
I was a bit vague by saying 'resources' - I meant specifically external scripts, and in my local version (not pushed yet I think) built-in scripts are included in the popup with a message informing the user that they will be saved when their scenes are.

I'm not sure what you mean about avoiding additional popups - I only add one popup asking to save all external scripts and listing them, and I don't think this can (or should) be avoided. Could you clarify this for me please?

File: editor/editor_node.cpp [hunk 4]
```diff
@@ -2945,34 +2945,86 @@ void EditorNode::_menu_option_confirm(int p_option, bool p_confirmed) {
...
+
+					save_confirmation->get_ok_button()->set_text(TTR("Save & Quit"));
+					save_confirmation->set_text((p_text) + unsaved_scripts);
+					save_confirmation->popup_centered();
+				}
+				break;
+			}
+
+			// User has chosen to save scripts before quit.
+			editor_data.get_editor("Script")->save_external_data();
```
[the remark this answers is not in the document]
Comment (ator-dev) [reply editor/editor_node.cpp:2982]: Good point, thank you - I forgot about that, I'll test it with the line removed.

**Comment by ator-dev** [comment]
By the way, I've been having a lot of trouble discarding scripts, which is why I haven't updated this PR with my updated version yet.

Part of my message to the contributors chat on the subject:

> I've been trying for over a week to make a method to discard a script (without closing the tab), for use in my save-or-discard-scripts prompt. This has turned out to be incredibly difficult; I've tried:
>
> - Set `text_editor`'s text to `text_file`'s text
> Problem: the `text_file` resource is updated with the unsaved text whenever you switch away from its tab, so only works for currently open tab.
>
> - Unref the script resource, then `set_edited_resource` to one newly loaded (current approach)
> Problem: again the cached one is retrieved, which contains the unsaved text and not the saved version.
>
> - Destroy the tab, then recreate and put in the same place with the same history
> Problem: horrible and brittle whichever way you approach it.
>
> - A bunch of variations of the above and other brief ideas which turned out to be infeasible.

> ... is it possible to load a resource without retrieving it from the cache? I'm using `ResourceLoader::load()` but it doesn't seem to have an option not to do that.

@KoBeWi helped me out with `ResourceLoader`, but none of the other options worked and it seems this PR is related: #63224 (thanks to @dzil123)

**Comment by ator-dev** [comment]
Start of another conversation relating to usability of the popup: https://chat.godotengine.org/channel/devel?msg=BABJmMxA8mX5sgFEu

**Comment by ator-dev** [comment]
Current state (discarding still not functional):
image

**Comment by Paulb23** [comment]
For popups I was thinking if we save the scene first that will in turn save the scrips, so we could avoid having this popup appear under that use case.

For `GDScript` looks like we are dependent on #63224 or do you mean `TextFiles` and `C#` are also not working?

As for `usability` we have this discussion from a while back #21959

**Comment by ator-dev** [comment]
> For popups I was thinking if we save the scene first that will in turn save the scrips, so we could avoid having this popup appear under that use case.

Isn't the purpose of this to show the user which script changes will be saved so that they can change their mind? For example, they might have temporarily changed or deleted some code and forgotten about it. Do you think this isn't valuable behaviour, and it's more important that scripts are simply always saved on quit (with scenes or with a dedicated prompt if there are no unsaved scenes)? Not criticising, I think I just misunderstood your intention.

> do you mean `TextFiles` and `C#` are also not working?

I added some behaviour for `TextFiles`, but haven't tested with either them or `C#` yet. At the moment I'm focusing on `GDScript` which I can't seem to discard; I agree we're dependent on that PR although a workaround could be added, like the 'do not save' flag I used before.

That usability discussion is interesting, I think it's a shame it was never picked up. Maybe some broader work is needed which would integrate improved script-saving behaviour with a more comprehensive scene-saving UI?

**Comment by Paulb23** [comment]
I guess it would depend if you see scenes as entirely separate entities to scripts. As a change in a script could be considered a change in the scene. Unless the script is not part of any scene. In which case it would be shown as a popup here. But I released that would only really apply to built-in scripts. So this approach does make more sense for the common usecase.

Personally not a fan of permanent temporary workarounds, would much rather wait for that other bug to get fixed first.

Yeah to get usability right would probably require some broader development in some way, shape, or form. I think for now what we have is good enough. Could chuck it in an `ItemList` if we're worried about users having that many unsaved scripts at a single time.

**Comment by ator-dev** [comment]
That's what I was thinking of, yeah - to me external scripts being saved when scenes are (which are entirely separate files that have references to scripts) is confusing to the user and could result in data loss if they messed around with their scripts in testing.

I agree, I don't want to add weird 'temporary' statefulness that will probably interact weirdly with other things and confuse contributors. The PR can probably wait until scripts have a nice way of being discarded.

I might look into using an `ItemList` - it's probably not a big deal but could improve readability, especially for large numbers of scripts. Definitely the current development is 'good enough' though and provides a good philosophy on which to build future work.

**Comment by KoBeWi** [comment]
I'm thinking about implementing a generic solution for all editor plugins:
https://github.com/godotengine/godot-proposals/issues/2153#issuecomment-1132753452
It would handle scripts, shaders and everything.

**Comment by akien-mga** [comment]
Superseded by #67503.
